crmd/notify.c

```diff
@@ -20,120 +20,501 @@
 #include <crm/crm.h>
 #include <crm/msg_xml.h>
 #include "notify.h"
+#include "crmd_messages.h"
 
 char *notify_script = NULL;
 char *notify_target = NULL;
+GListPtr notify_list = NULL;
+
+typedef struct {
+    char *id;
+    char *path;
+    int timeout;
+    char *tstamp_format;
+    char *info;
+    char *recipient;
+} notify_entry_t;
+
+enum notify_keys_e{
+    CRM_notify_recipient = 0,
+    CRM_notify_node,
+    CRM_notify_nodeid,
+    CRM_notify_rsc,
+    CRM_notify_task,
+    CRM_notify_interval,
+    CRM_notify_desc,
+    CRM_notify_status,
+    CRM_notify_target_rc,
+    CRM_notify_rc,
+    CRM_notify_kind,
+    CRM_notify_version,
+    CRM_notify_node_sequence,
+    CRM_notify_timestamp,
+    CRM_notify_info,
+};
+
+/*
+ * to allow script compatibility we can have more than one
+ * set of environment variables
+ */
+
+static const char *notify_keys[][3] = 
+{
+    [CRM_notify_recipient]     = {"CRM_notify_recipient", NULL},
+    [CRM_notify_node]          = {"CRM_notify_node", NULL},
+    [CRM_notify_nodeid]        = {"CRM_notify_nodeid", NULL},
+    [CRM_notify_rsc]           = {"CRM_notify_rsc", NULL},
+    [CRM_notify_task]          = {"CRM_notify_task", NULL},
+    [CRM_notify_interval]      = {"CRM_notify_interval", NULL},
+    [CRM_notify_desc]          = {"CRM_notify_desc", NULL},
+    [CRM_notify_status]        = {"CRM_notify_status", NULL},
+    [CRM_notify_target_rc]     = {"CRM_notify_target_rc", NULL},
+    [CRM_notify_rc]            = {"CRM_notify_rc", NULL},
+    [CRM_notify_kind]          = {"CRM_notify_kind", NULL},
+    [CRM_notify_version]       = {"CRM_notify_version", NULL},
+    [CRM_notify_node_sequence] = {"CRM_notify_node_sequence", NULL},
+    [CRM_notify_timestamp]     = {"CRM_notify_timestamp", NULL},
+    [CRM_notify_info]          = {"CRM_notify_info", NULL}
+};
 
+/*
+ * higher accuracy time stuff to be generalized and moved to 
+ * e.g. lib/common/utils.c|iso8601.c
+ */
+
+#include <time.h>
+#include <ctype.h>
+#include <crm/common/iso8601.h>
+
+typedef struct crm_time_us crm_time_hr_t;
+crm_time_hr_t *crm_time_hr_convert(crm_time_hr_t *target, crm_time_t *dt);
+void crm_time_set_hr_dt(crm_time_t *target, crm_time_hr_t *hr_dt);
+crm_time_hr_t *crm_time_timeval_hr_convert(crm_time_hr_t *target, 
+                                           struct timeval *tv);
+crm_time_hr_t *crm_time_hr_new(const char *date_time);
+void crm_time_hr_free(crm_time_hr_t * hr_dt);
+char *crm_time_format_hr(const char *format, crm_time_hr_t * hr_dt);
+
+crm_time_t *parse_date(const char *date_str); /* in iso8601.c global but
+                                                 not in header */
+
+
+struct crm_time_us {
+    int years;
+    int months;                 /* Only for durations */
+    int days;
+    int seconds;
+    int offset;                 /* Seconds */
+    bool duration;
+    int useconds;
+};
```

Comment:
Why not just add useconds to crm_time_s? The existing functions could continue to ignore it, and just the new functions would use it. We can add fields to the end of structs without breaking binary compatibility.